### PR TITLE
code and config files

### PR DESCRIPTION
**Pull request by LittleDinoC**
Add code and config files of the main lab.

### CODE REVIEW

**Comment by gongchuanyang** [comment]
你好，同学，src/generate.py中的第650处的代码好像有点问题，  att = att[tl:tr] 中的tl, tr 是否有问题，  att的shape 为  （atten_tmp.shape[0], input_length）， 而att = att[tl:tr] 并不能索引到tokens，tl, tr 这个是tokens(合并后的单词)的索引，而att 里面是分词后的token,  并不匹配， 请指教

tl, tr = 0, len(tokens) 
curr_st = len(tokens) - len(curr_tokens)  
attns = []
for r in range_:
            att = torch.zeros(atten_tmp.shape[0], input_length)
            for i in range(r[0], r[1] + 1):
                att += atten_tmp[:, i]
            att /= (r[1] - r[0] + 1)
            att = torch.mean(att, dim=0)
            att = att[tl:tr]
            if att.shape[0] > 1:
                att = att / sum(att[1:]).item()
            attns.append(att)

**Comment by oneal2000** [comment]
感谢你指出的问题！ 我核实了一下，generate.py是我们老版本的代码，这个版本的代码确实有这一问题，导致Query Formulation的时候，在某些情况（有合并token操作的情况）会发生错位。

我会在本月底之前，尽快把最新版本的代码放上来，这一版本没有这个问题。

更新代码后，我们会有以下行动：
（1）在主页声明，这一版本代码的具体问题
（2）将我们最新版的代码同步到github
（3）开放 Issue，欢迎所有人指出任何问题！

**Comment by gongchuanyang** [comment]
你好作者，还有一个疑问，
        real_pairs = []
        for i in range(len(tokens)):
            tok, att = tokens[i], forward_attns[i]  # 这个地方好像没有把出现幻觉的token排除，在下面的排序过程中，好像依然会保留幻觉token，导致检索的文档也会出现错误
            if match(tok):
                real_pairs.append((att, tok, i))
        
        if "retrieve_keep_top_k" in self.__dict__:
            top_k = min(self.retrieve_keep_top_k, len(real_pairs))
        elif "retrieve_keep_ratio" in self.__dict__:
            top_k = int(len(real_pairs) * self.retrieve_keep_ratio)
        
        real_pairs = sorted(real_pairs, key = lambda x:x[0])
        real_pairs = real_pairs[:top_k]
        real_pairs = sorted(real_pairs, key = lambda x:x[2])
        return " ".join([x[1] for x in real_pairs])

**Comment by gongchuanyang** [comment]
你好，作者，能否提供一份新的generate.py文件, 1362444584@qq.com，感谢，你的工作对我很有启发意义

**Comment by oneal2000** [comment]
> 你好作者，还有一个疑问， real_pairs = [] for i in range(len(tokens)): tok, att = tokens[i], forward_attns[i] # 这个地方好像没有把出现幻觉的token排除，在下面的排序过程中，好像依然会保留幻觉token，导致检索的文档也会出现错误 if match(tok): real_pairs.append((att, tok, i))

对于“是否选择Uncertain Token”，你可以把它理解成：这其实是一个可选的超参。选择这个 Token 其实有好处也有坏处

好处是：这个token其实可以指导搜索，比如这个token是地名，那我们的 Relevant Passage的主题也大概率是相同类别的；并且Uncertain Token也不一定是错的

坏处是：如果Uncertain Token是错的，那可能会引入一定噪音。

感谢您的提问！我们后续可能做消融实验和Case study，来探究这个超参的影响

也真心感谢您对我们项目的关注！
